### DISCUSSION TITLE
Getting remote foreground process name over an SSH connection

### DISCUSSION THREAD

**Discussion by mrjones2014** [kovidgoyal/kitty#6229] [Q&A]
For context: https://github.com/mrjones2014/smart-splits.nvim/issues/103

To support the plugin, I have a Kitten which checks the foreground process name to check if vim or nvim is running:

```python
# vim_id is a pattern, defaults to "n?vim"
def is_window_vim(window, vim_id):
    from kittens.tui.loop import debug
    debug(vim_id)
    fp = window.child.foreground_processes
    return any(re.search(vim_id, p['cmdline'][0] if len(p['cmdline']) else '', re.I) for p in fp)
```

However this does not report the remote process name over SSH. Wezterm has the same limitation. It appears that only tmux does this.

Is there any way to work around this?

Also, is there full documentation somewhere on the Python API? I couldn't find it on the main docs website.

**Comment by kovidgoyal** [comment]
Assuming the user is using the ssh kitten you can ssh into the box using
the existing ssh master connection and use ps or similar to find the process. 

Or just check if the window title contains nvim. 

And no, the Python API is not documented anywhere, its not guaranteed to
be stable.

**Comment by gzagatti** [comment]
Checking whether the title window contains nvim, works partially. If the user is located in a folder containing the word vim the matching on window title fails. Let's say how do you differentiate a window title `/home/linuxbrew/.linuxbrew/Cellar/vim` from `/home/linuxbrew/.linuxbrew/bin/vim`. The former is visitng the folder `vim` and the latter running the program `vim`.

Alternatively, it would be great if the name of the running process was added to the window object when updating the title of the window when using ssh kitten? That would help with reducing ambiguity.

Alternatively, would it be possible to elaborate on. How do you access the ssh master connection? How do you discover the pid of the process running under present window?

> Assuming the user is using the ssh kitten you can ssh into the box using the existing ssh master connection and use ps or similar to find the process.

**Comment by kovidgoyal** [comment]
On Thu, May 11, 2023 at 02:27:47AM -0700, Guilherme Zagatti wrote:
> Checking whether the title window contains nvim, works partially. If the user is located in a folder containing the word vim the matching on window title fails. Let's say how do you differentiate a window title `/home/linuxbrew/.linuxbrew/Cellar/vim` from `/home/linuxbrew/.linuxbrew/bin/vim`. The former is visitng the folder `vim` and the latter running the program `vim`.

You match on the *whole word* nvim. not the substring vim. nvim's
default title template for example is nvim %filename IIRC. So look for
titles matching nvim\s+\S+

Of course this relies on users not changing the title template.

> 
> Alternatively, it would be great if the name of the running process was added to the window object when updating the title of the window when using ssh kitten? That would help with reducing ambiguity.

There is no protocol for reporting anything other than titles over the
tty. I could of course invent one, but then nothing would be using it,
so it wouldnt help you.

> 
> Alternatively, would it be possible to elaborate on. How do you access the ssh master connection? How do you discover the pid of the process running under present window?

The ssh kitten uses a controlmaster to connect (unless the user has
configured it not too). In a kitten you can check the commnadline of the
foregound process of the window you are interested in. If you detect its
the ssh kitten (and there are helper functions in the code base to do
that), then you can extract the path to the control master from the
command line. Then using that you run ssh <control master flags>
your_script kitty_window_id

where your script will go through /proc/pid/environ looking for a
process that has the env var KITTY_WINDOW_ID=kitty_window_id. The
process group of this process will be the one running in the window
of interest.

That said a better approach to this problem is if you can modify the
nvim environment on the remote host. Then you can simply have the user
set an appropriate title template that you can detect, or have add some
lines to nvimrc that transmit data you are interested in over the tty
using escape codes. IIRC iTerm has some protocol that allows setting of
arbitrary variables received over the tty, I could implement that.

-- 
_____________________________________

Dr. Kovid Goyal 
https://www.kovidgoyal.net
https://calibre-ebook.com
_____________________________________

**Comment by gzagatti** [comment]
Thanks for the quick response, as always really helpful.

I went with the first option which is simpler but less robust:

```python
from kittens.ssh.utils import is_kitten_cmdline

def is_window_vim(boss, window, vim_id):
    fp = window.child.foreground_processes
    wtitle = window.child_title
    for p in fp:
        q = list(p['cmdline'] or ())
        if is_kitten_cmdline(q):
            wtitle = wtitle.split(": ")[1]
            vim_id = f"^{vim_id}(\s+\S+|$)"
            if re.search(vim_id, wtitle, re.I):
                return True
            else:
                return False
        else:
            if len(q) == 0:
                continue
            if re.search(vim_id, p['cmdline'][0], re.I):
                return True
    return False
```

The third option is interesting, but it is quite complex and I am worried it could be slow.

> There is no protocol for reporting anything other than titles over the tty.

Shame that there are no protocols for reporting anything other than titles. I guess an alternative would be to transmit all the useful information (e.g. label, working directory, running process, hostname) over the title protocol encoded as a string, then have kitty only show the label in the window title. However, this might be not be the desired use of the title.

> That said a better approach to this problem is if you can modify the nvim environment on the remote host.

The problem is not so much with nvim but with overloading the `ctrl-h`, `ctrl-j`, `ctrl-k` and `ctr-l` keys. Since they are first captured by `kitty` we need to pass them to nvim if that was the calling process, otherwise we let kitty handle them.

**Comment by mrjones2014** [comment]
> Of course this relies on users not changing the title template.

I don't really see title matching as a real solution tbh. Too much inconsistency and too much unknowns, it's not really reliable.

> There is no protocol for reporting anything other than titles over the
tty.

How does `tmux` do it? In tmux we just use this snippet to detect vim/nvim and it works over SSH.

```tmux
is_vim="ps -o state= -o comm= -t '#{pane_tty}' \
    | grep -iqE '^[^TXZ ]+ +(\\S+\\/)?g?(view|n?vim?x?)(diff)?$'"
```

> Assuming the user is using the ssh kitten you can ssh into the box using the existing ssh master connection and use ps or similar to find the process.

This can't be guaranteed either.

FWIW, we worked around this in Wezterm using [user vars](https://wezfurlong.org/wezterm/recipes/passing-data.html#user-vars) -- we just set `IS_NVIM=true` when nvim loads, and set it back to false on `ExitPre` autocmd. Since this is an escape code interpreted directly by the terminal, it works no differently over an SSH connection than locally. Kitty doesn't have anything like that, though, right?

**Reply by kovidgoyal** [reply]
See my last reply, user vars is precisely the iterm arbitrary variables I was talking about. And no they aren't implemented in kitty, though it should be a trivial patch to implement them.

**Comment by kovidgoyal** [comment]
And added support for SetUserVar. Note that ExitPre is not sufficient, you need to handle suspend as well. Not to mention that exitpre can be cancelled, VimLeavePre is the correct event.

**Reply by mrjones2014** [reply]
Any docs on `SetUserVar`, or is it essentially the same as how it works in Wezterm?

**Comment by kovidgoyal** [comment]
It's the same as it works on iTerm which originated it. I would assume
wezterm works the same way.
